File: orgo_mcp/tools/agent.py
"""AI agent completions tool — the core agent API.

Uses the OpenAI-compatible POST /api/v1/chat/completions endpoint.
The AI agent handles the full loop: screenshot -> decide -> act -> repeat.
"""

import json

import httpx

from orgo_mcp.server import mcp
from orgo_mcp.auth import get_current_api_key, resolve_computer_id
from orgo_mcp.client import ORGO_V1_BASE
from orgo_mcp.errors import handle_orgo_error
from orgo_mcp.models import CompletionsInput


@mcp.tool(
    name="orgo_completions",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": False, "openWorldHint": True},
)
async def orgo_completions(params: CompletionsInput) -> str:
    """Send an instruction to an AI agent controlling an Orgo computer.

    The agent sees the screen, clicks, types, and runs commands autonomously until
    the task is done. Returns the agent's final response and a thread_id for follow-up.
    Use thread_id to continue multi-turn conversations.

    Models: claude-sonnet-4.6 (fast, default), claude-opus-4.6 (most capable).
    Set anthropic_key for BYOK mode (your key, no Orgo credits consumed).
    """
    try:
        api_key = get_current_api_key(mcp)
        computer_id = resolve_computer_id(params.computer_id)

        body = {
            "model": params.model,
            "messages": [{"role": "user", "content": params.instruction}],
            "computer_id": computer_id,
        }
        if params.thread_id:
            body["thread_id"] = params.thread_id
        if params.max_steps is not None:
            body["max_steps"] = params.max_steps

        headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        }
        if params.anthropic_key:
            headers["X-Anthropic-Key"] = params.anthropic_key

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{ORGO_V1_BASE}/chat/completions",
                json=body,
                headers=headers,
                timeout=300.0,
            )
            response.raise_for_status()
            return json.dumps(response.json(), indent=2)
    except Exception as e:
        return handle_orgo_error(e)

File: src/tools/agent.ts
/**
 * AI agent completions tool -- the core agent API.
 *
 * Uses the OpenAI-compatible POST /api/v1/chat/completions endpoint.
 * The AI agent handles the full loop: screenshot -> decide -> act -> repeat.
 */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { getApiKey, resolveComputerId } from "../auth.js";
import { ORGO_V1_BASE } from "../client.js";
import { handleError } from "../errors.js";
import { HttpError } from "../errors.js";

export function registerAgentTools(server: McpServer): void {
  server.tool(
    "orgo_completions",
    "Send an instruction to an AI agent controlling an Orgo computer. The agent sees the screen, clicks, types, and runs commands autonomously until the task is done. Returns the agent's final response and a thread_id for follow-up.",
    {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
      instruction: z.string().min(1).describe("What the AI agent should do (natural language)"),
      model: z.string().default("claude-sonnet-4.6").describe("Model: claude-sonnet-4.6 or claude-opus-4.6"),
      thread_id: z.string().optional().describe("Thread ID to continue a previous conversation"),
      max_steps: z.number().int().min(1).max(500).optional().describe("Max agent steps (default: 100)"),
      anthropic_key: z.string().optional().describe("Your Anthropic API key for BYOK mode (no Orgo credits used)"),
    },
    async ({ computer_id, instruction, model, thread_id, max_steps, anthropic_key }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);

        const body: Record<string, unknown> = {
          model,
          messages: [{ role: "user", content: instruction }],
          computer_id: id,
        };
        if (thread_id) body.thread_id = thread_id;
        if (max_steps !== undefined) body.max_steps = max_steps;

        const headers: Record<string, string> = {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        };
        if (anthropic_key) headers["X-Anthropic-Key"] = anthropic_key;

        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), 300000);

        try {
          const response = await fetch(`${ORGO_V1_BASE}/chat/completions`, {
            method: "POST",
            headers,
            body: JSON.stringify(body),
            signal: controller.signal,
          });

          if (!response.ok) {
            let detail = "";
            try {
              const errJson = (await response.json()) as Record<string, unknown>;
              detail = (errJson.error as string) || "";
            } catch {
              // ignore
            }
            throw new HttpError(response.status, detail, response.statusText);
          }

          const data = await response.json();
          return { content: [{ type: "text" as const, text: JSON.stringify(data, null, 2) }] };
        } finally {
          clearTimeout(timer);
        }
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    }
  );
}
